### PR TITLE
Implemented Collection::zip() and zipWith()

### PR DESCRIPTION
**Pull request by lorenzo**
Another common iteration method that I figured was missed was `zip()`. I also implemented `zipWith()` since I think it is quite handy.

Edit:

Also took the liberty of implementing something I keep doing myself in some projects: `skip()` and `last()`

### CODE REVIEW

File: src/Collection/CollectionInterface.php [hunk 1]
```diff
@@ -527,6 +536,14 @@ public function firstMatch(array $conditions);
      * @return mixed The first value in the collection will be returned.
      */
     public function first();
+
+    /**
+     * Returns the last result in this collection
+     *
+     * @return mixed The first value in the collection will be returned.
```
Comment (ceeram) [inline src/Collection/CollectionInterface.php]: first should be last here

File: src/Collection/CollectionInterface.php [hunk 2]
```diff
@@ -832,6 +849,44 @@ public function unfold(callable $transformer = null);
...
+     * Combines the elements of this collection with each of the elements of the
+     * passed iterables, using their positional index as a reference.
+     *
+     * The resulting element will be the return value of the $callable function.
+     *
+     * ### Example:
+     *
+     * ```
+     * $collection = new Collection([1, 2]);
+     * $zipped = $collection->zip([3, 4], [5, 6], function () {
```
Comment (ceeram) [inline src/Collection/CollectionInterface.php]: should be zipWith()

File: src/Collection/CollectionInterface.php [hunk 3]
```diff
@@ -483,6 +483,15 @@ public function sample($size = 10);
     public function take($size = 1, $from = 0);
 
     /**
+     * Returns a new collection that will the specified amount of elements.
```
Comment (ceeram) [inline src/Collection/CollectionInterface.php]: missing skip in this sentence

File: src/Collection/CollectionInterface.php [hunk 4]
```diff
@@ -832,6 +849,44 @@ public function unfold(callable $transformer = null);
...
+     * passed iterables, using their positional index as a reference.
+     *
+     * ### Example:
+     *
+     * ```
+     * $collection = new Collection([1, 2]);
+     * $collection->zip([3, 4], [5, 6])->toList(); // returns [[1, 3, 5], [2, 4, 6]]
+     * ```
+     *
+     * @param array|\Traversable ...$items The collections to zip.
```
Comment (ceeram) [inline src/Collection/CollectionInterface.php]: leading dots for the argument in docblock are intended?

File: src/Collection/Iterator/ZipIterator.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * You can also chose a custom function to zip the elements together, such
+ * as doing a sum by index:
+ *
+ * ### Example
+ * {{{
+ *  $iterator = new ZipIterator([[1, 2], [3, 4]], function ($a, $b) {
+ *    return $a + $b;
+ *  });
+ *  $iterator->toList(); // Returns [3, 6]
```
Comment (jadb) [inline src/Collection/Iterator/ZipIterator.php]: Should return `[4, 6]`.

**Comment by jadb** [comment]
I like! :+1: 

Why `zip` and not `group`? We'd end up with `group()`, `groupBy()`, `groupWith()`?

File: src/Collection/Iterator/ZipIterator.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+use Cake\Collection\CollectionInterface;
+use Cake\Collection\CollectionTrait;
+use MultipleIterator;
+
+/**
+ * Creates an iterator that returns elements grouped in pairs
+ *
+ * ### Example
+ *
+ * {{{
```
Comment (antograssiot) [inline src/Collection/Iterator/ZipIterator.php]: you're using the old code fence in tm this file for code blocks.

**Comment by lorenzo** [comment]
@jadb `zip` is the standard name for what this does, python and haskell use it, for example. Also group is not a good choice since the Query object already has that method and I'd like all collection methods to be  distinct of those you can use in the query.

File: src/Collection/CollectionInterface.php (hunk 4 above)
Comment (lorenzo) [inline src/Collection/CollectionInterface.php]: Yes, I'm starting to use the php 5.6 way of hinting variable arguments of the same type

File: src/Collection/Iterator/ZipIterator.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @return mixed
+     */
+    public function current()
+    {
+        if ($this->_callback === null) {
+            return parent::current();
+        }
+
+        return call_user_func_array($this->_callback, parent::current());
```
Comment (markstory) [inline src/Collection/Iterator/ZipIterator.php]: Neat, nice that `MultipleIterator` made the hard part of this easier.

**Comment by markstory** [comment]
Awesome work @lorenzo

File: src/Collection/Iterator/ZipIterator.php (hunk 7 above)
Comment (lorenzo) [inline src/Collection/Iterator/ZipIterator.php]: Yeah, It simplified a lot my first draft code :D

**Pull request merged**